### ISSUE TITLE
Versioning

### ISSUE THREAD

**Issue by jwilm** [phildawes/racer#619]
I was about to make a cut of Racer this morning and realized that we've changed a public API in an incompatible way. Specifically, we landed #498 which removed some unsafe code related to the `FileCache`. This was a very welcome patch of course, but it means we can't release from master without bumping the major version.

Options:
1. **Revert the FileCache change**. This is probably no longer easy to do since there's been quite a few changes to master.
2. **Actually cut a 2.0**. This isn't exactly great since there's a lot of other API work that could be done, but it would allow us to move forward with minimal effort.
3. **Add a FileCache shim**. This could expose an API matching the old FileCache, but it's pure technical debt. The shim would be thrown out when we actually

I'm inclined to just cut a version 2.0. This is the simplest option, there's just one type with breaking changes (so upgrading is easy), and there's really no limit to how many major versions we can have.

There's a bunch of fixes on master that really ought to be released, and there's more on the way. Assuming we release a 2.0 now, I would want to figure out what a minimal API we can commit to for 3.0 would look like and hold off on breaking changes until that's decided. Until then, the focus would be on backwards compatible changes like increasing stability and making racer work for more cases.

@phildawes, thoughts?

**Comment by phildawes** [comment]
Hello! Does anything other than racerd use this api? I think the other plugins just use the command line interface. If that's true, then probably best to cut a 1.3 and not worry about it too much.
(I'm happy for a 2.0 release if you think it's warranted though)

**Comment by jwilm** [comment]
Interesting thought! I think most integrations are using either the CLI, stdio pipe, or racerd. Thanks, Phil. 

@nrc, does RLS use the FileCache API?

**Comment by nrc** [comment]
Not really. Our entry point to Racer is `complete_from_file`. We do create an empty `FileCache`, but only to make a `Session` which we pass to `complete_from_file`, we don't use any functionality beyond that.

**Comment by jwilm** [comment]
@nrc thanks for the input

After thinking on this a bit (and reading comments in [futures-rs#160]), I think we should go for a 2.0 just to avoid any unexpected breakage. With that, I'd like to reduce the public API to a minimal set of things needed to make completions and etc; hopefully doing so will make this less likely to happen again.

[futures-rs#160]: https://github.com/alexcrichton/futures-rs/issues/160

**Comment by jwilm** [comment]
I've got most of the 2.0 API worked out at this point, and I'll hopefully be submitting a patch for review tomorrow. The command line and daemon APIs will remain unchanged, and there are minor changes to how Rust code integrates with Racer (simpler). The main thing to look forward to in the patch will be a very narrow public API which will allow us to iterate quickly on Racer internals. There's also a number of fixes on master that will become available once we can publish to crates.io without inadvertently breaking anyone's code :).
